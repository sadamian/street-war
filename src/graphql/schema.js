import {gql} from 'apollo-server-express';


export const resolvers = {
  Query: {
    // @ts-ignore
    users: async (_, {}, { dataSources }) => await dataSources.userAPI.getUsers(),
  },
};

export const typeDefs = gql`

  type User {
    _id: ID!
    name: String!
    secret: String!
    admin: Boolean
  }
  
  type Query {
    users: [User]
  }
  
`;
