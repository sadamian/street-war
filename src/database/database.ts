import mongoose from "mongoose";

let database: mongoose.Connection;

export const connectBDD = (dbUri: string) => {
  const uri = dbUri;
  if (database) {
    return;
  }
  mongoose.connect(uri, {
    useNewUrlParser: true,
    useFindAndModify: true,
    useUnifiedTopology: true,
    useCreateIndex: true,
  });
  database = mongoose.connection;
  database.once("open", async () => {
    console.log("Connected to database");
  });
  database.on("error", () => {
    console.log("Error connecting to database");
  });
};

export const disconnectBDD = () => {
  if (!database) {
    return;
  }
  mongoose.disconnect();
};
