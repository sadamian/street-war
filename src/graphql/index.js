import { ApolloServer } from "apollo-server-express";

import UserAPI from "./datasources/users";
import { resolvers, typeDefs } from "./schema";

export const dataSources = () => ({
  userAPI: new UserAPI(),
});

const server = new ApolloServer({
  typeDefs,
  resolvers,
  dataSources,
});

export default server;
