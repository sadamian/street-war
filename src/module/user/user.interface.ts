import { UUID } from "../../common/type/uuid.type";

export interface IUser extends IPublicUser {
  password: string;
}

export interface IPublicUser {
  _id?: UUID;
  secret: string;
  name: string;
  admin?: boolean;
}
