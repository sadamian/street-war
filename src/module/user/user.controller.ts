import { Request, Response } from "express";
import { IPublicUser, IUser } from "./user.interface";
import {
  findAll,
  create,
  getUserInfo,
  find,
} from "./user.database/user.database.service";
import { UserModel } from "./user.database/user.database.model";

/**
 * Controller Definitions
 */

export async function getAllUserController(req: Request, res: Response) {
  try {
    const users: IUser[] = await findAll();
    res.status(200).send(users);
  } catch (e) {
    res.status(404).send(e.message);
  }
}

export async function createUserController(req: Request, res: Response) {
  try {
    const item: IUser = req.body;
    const user: IPublicUser = await create(item);
    res.status(201).send(user);
  } catch (e) {
    res.status(500).send(e.message);
  }
}

export async function getUserController(req: Request, res: Response) {
  try {
    const user_id: string = req.params?.user_id;
    const user: IPublicUser = await find(user_id);
    res.status(201).send(user);
  } catch (e) {
    res.status(500).send(e.message);
  }
}

export async function userInfoController(req: Request, res: Response) {
  try {
    // switch to auth
    if (req.headers["user-id"] && typeof req.headers["user-id"] === "string") {
      const user_id: string = req.headers["user-id"];
      const info = await getUserInfo(user_id);
      res.status(201).send(info);
    } else {
      throw new Error("no token");
    }
  } catch (e) {
    res.status(500).send(e.message);
  }
}
