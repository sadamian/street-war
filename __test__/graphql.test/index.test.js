import { ApolloServer } from "apollo-server-express";
import { createTestClient } from "apollo-server-testing";
import gql from "graphql-tag";
const { setupDB } = require("../bdd-test.setup");

import UserAPI from "../../src/graphql/datasources/users";
import { typeDefs, resolvers } from "../../src/graphql/schema";
import { UserModel } from "../../src/module/user/user.database/user.database.model";
import { users } from "../fixture/user.fixture";
import { IUser } from "../../src/module/user/user.interface";

const userAPI = new UserAPI();

// create a test server to test against, using our production typeDefs,
// resolvers, and dataSources.
const server = new ApolloServer({
  typeDefs,
  resolvers,
  dataSources: () => ({ userAPI }),
});
const { query } = createTestClient(server);

describe("GrapQL:Users", () => {
  setupDB("test-user-graphql-bdd");

  beforeAll(async () => {
    await UserModel.insertMany(users);
  });

  afterAll(async () => {
    await UserModel.remove({});
  });

  it("Users : Get only name", async () => {
    const GET_USERS = gql`
      query userList {
        users {
          name
        }
      }
    `;

    userAPI.getUsers = jest.fn(() => users.map((u) => ({ name: u.name })));
    const res = await query({ query: GET_USERS });
    expect(res).toMatchSnapshot();
  });

  it("Users : Get name and _id", async () => {
    const GET_USERS = gql`
      query userList {
        users {
          name
          _id
        }
      }
    `;
    userAPI.getUsers = jest.fn(() =>
      users.map((u) => ({ name: u.name, _id: u._id }))
    );
    const res = await query({ query: GET_USERS });
    expect(res).toMatchSnapshot();
  });
});
