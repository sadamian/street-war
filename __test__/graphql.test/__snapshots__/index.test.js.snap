// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GrapQL:Users Users : Get name and _id 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "_id": "603ec467770f507d7910580a",
        "name": "Proutios",
      },
      Object {
        "_id": "603ec467770f507d7910580b",
        "name": "Proutini",
      },
      Object {
        "_id": "603ec467770f507d7910580c",
        "name": "Proutalop",
      },
      Object {
        "_id": "603ec467770f507d7910580d",
        "name": "Daolika",
      },
      Object {
        "_id": "603ec467770f507d7910580e",
        "name": "Ronuni",
      },
      Object {
        "_id": "603ec467770f507d7910580f",
        "name": "Louika",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`GrapQL:Users Users : Get only name 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "name": "Proutios",
      },
      Object {
        "name": "Proutini",
      },
      Object {
        "name": "Proutalop",
      },
      Object {
        "name": "Daolika",
      },
      Object {
        "name": "Ronuni",
      },
      Object {
        "name": "Louika",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
